### PR TITLE
WordPress 5.8 Compatibility: Add support for `block_editor_settings_all`

### PR DESCRIPTION
**Pull request by sdixon194**
The filter `block_editor_settings` is [being deprecated in favor of `block_editor_settings_all`](https://core.trac.wordpress.org/ticket/52920) for WordPress 5.8.

#### Changes proposed in this Pull Request:

* Adds support for block_editor_settings_all when running WordPress 5.8 or higher. 

#### Jetpack product discussion

https://github.com/Automattic/jetpack/issues/19654

#### Does this pull request change what data or activity we track or use?

No.
#### Testing instructions:

* Smoke test using the block editor.
* Enable the option allowing you to copy posts under Jetpack -> Writing -> Compose settings.
* Make sure the copy post option still works.

### CODE REVIEW

**Review by anomiex — APPROVED** [review]
Looks sensible to me. Haven't tested though.

File: projects/plugins/jetpack/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php [hunk 1]
```diff
@@ -64,11 +65,15 @@ public function init_actions() {
 
 		require_once __DIR__ . '/functions.editor-type.php';
 		add_action( 'edit_form_top', 'Jetpack\EditorType\remember_classic_editor' );
-		add_filter( 'block_editor_settings', 'Jetpack\EditorType\remember_block_editor', 10, 2 );
 		add_action( 'login_init', array( $this, 'allow_block_editor_login' ), 1 );
 		add_action( 'enqueue_block_editor_assets', array( $this, 'enqueue_block_editor_assets' ), 9 );
 		add_action( 'enqueue_block_assets', array( $this, 'enqueue_block_assets' ) );
 		add_filter( 'mce_external_plugins', array( $this, 'add_tinymce_plugins' ) );
+		if ( version_compare( $wp_version, '5.8', '>=' ) ) {
+			add_filter( 'block_editor_settings_all', 'Jetpack\EditorType\remember_block_editor', 10, 2 );
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php:73]: I don't think that's going to work as well in WordPress 5.8.

In WordPress 5.7, the `block_editor_settings` filter's second parameter was an instance of `WP_Post`. We use it here:
https://github.com/Automattic/jetpack/blob/26db3e436ccca3e16a62d02d95e792a81f38a1e4/projects/plugins/jetpack/modules/wpcom-block-editor/functions.editor-type.php#L32

In WordPress 5.8, the `block_editor_settings_all` filter's second parameter is different: it is now an instance of `WP_Block_Editor_Context`.

**Review by jeherve — CHANGES_REQUESTED** [review]
I left a small remark below. 

I also think it would be nice to add a comment so we can easily find that backwards compatibility check later and remove it once WP 5.8 is the minimum supported version.

Maybe something like this?

```diff
diff --git a/projects/plugins/jetpack/modules/copy-post.php b/projects/plugins/jetpack/modules/copy-post.php
index 069adef24c..dc92af45a1 100644
--- a/projects/plugins/jetpack/modules/copy-post.php
+++ b/projects/plugins/jetpack/modules/copy-post.php
@@ -70,7 +70,10 @@ class Jetpack_Copy_Post {
 		add_filter( 'default_content', array( $this, 'filter_content' ), 10, 2 );
 		add_filter( 'default_excerpt', array( $this, 'filter_excerpt' ), 10, 2 );
 
-		// Required to avoid the block editor from adding default blocks according to post format.
+		/*
+		 * Required to avoid the block editor from adding default blocks according to post format.
+		 * @todo: simplify once WordPress 5.8 is the minimum required version.
+		 */
 		if ( version_compare( $wp_version, '5.8', '>=' ) ) {
 			add_filter( 'block_editor_settings_all', array( $this, 'remove_post_format_template' ) );
 		} else {
diff --git a/projects/plugins/jetpack/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php b/projects/plugins/jetpack/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php
index db35050d99..7d89bb0052 100644
--- a/projects/plugins/jetpack/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php
+++ b/projects/plugins/jetpack/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php
@@ -69,6 +69,8 @@ class Jetpack_WPCOM_Block_Editor {
 		add_action( 'enqueue_block_editor_assets', array( $this, 'enqueue_block_editor_assets' ), 9 );
 		add_action( 'enqueue_block_assets', array( $this, 'enqueue_block_assets' ) );
 		add_filter( 'mce_external_plugins', array( $this, 'add_tinymce_plugins' ) );
+
+		// @todo: simplify once WordPress 5.8 is the minimum required version.
 		if ( version_compare( $wp_version, '5.8', '>=' ) ) {
 			add_filter( 'block_editor_settings_all', 'Jetpack\EditorType\remember_block_editor', 10, 2 );
 		} else {
diff --git a/projects/plugins/jetpack/modules/wpcom-block-editor/functions.editor-type.php b/projects/plugins/jetpack/modules/wpcom-block-editor/functions.editor-type.php
index ea053954e3..40d76be7a4 100644
--- a/projects/plugins/jetpack/modules/wpcom-block-editor/functions.editor-type.php
+++ b/projects/plugins/jetpack/modules/wpcom-block-editor/functions.editor-type.php
@@ -24,11 +24,21 @@ function remember_classic_editor( $post ) {
 /**
  * Remember when the block editor was used to edit a post.
  *
- * @param  object $editor_settings This is hooked into a filter and this is the settings that are passed in.
- * @param  object $post            The post being editted.
- * @return object                  The unmodified $editor_settings parameter.
+ * @todo: simplify once WordPress 5.8 is the minimum required version.
+ *
+ * @param  array                           $editor_settings This is hooked into a filter and this is the settings that are passed in.
+ * @param  WP_Post|WP_Block_Editor_Context $post            In WP 5.7- the post being editted. In WP 5.8+, the block editor context.
+ *
+ * @return array The unmodified $editor_settings parameter.
  */
 function remember_block_editor( $editor_settings, $post ) {
+	if (
+		is_a( $post, 'WP_Block_Editor_Context' )
+		&& ! empty( $post->post )
+	) {
+		$post = $post->post;
+	}
+
 	$post_type = get_post_type( $post );
 
 	if ( $post_type && can_edit_post_type( $post_type ) ) {

```

**Review by kraftbj — APPROVED** [review]

**Comment by sdixon194** [comment]
Cherry picked to release branch in 1bb54d22efb441cfcc5f8626add25267fcd430cc

**Comment by sdixon194** [comment]
Deployed to wpcom in r228838

**Pull request merged**